### PR TITLE
Install sktm via pip, and switch to use tox

### PR DESCRIPTION
**Pull request by danghai**
I make PR to install sktm via pip, and switch to run tests, flake8, and pylint via tox

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+[testenv]
+passenv = TRAVIS TRAVIS_*
+deps =
+    coverage
+    mock
+    coveralls
+    pytest
+commands =
+    coverage run --source=sktm,tests -m pytest -v -s .
+    coverage report -m --omit="*/tests*"
```
Comment (veruu) [inline tox.ini]: Instead of running coverage for tests and then removing it from the report, it would be easier to just not run it for tests directory. You can see how it's done in skt

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,32 @@
+
```
Comment (veruu) [inline tox.ini]: This empty line is useless here, can you remove it?

File: .travis.yml [hunk 3]
```diff
@@ -1,16 +1,13 @@
 language: python
 python:
     - "2.7"
+env:
+    - TOX_ENV=py27,flake8,pylint
 install:
     - pip install -U pip
     - pip install -r requirements.txt
     - pip install -r test-requirements.txt
```
Comment (veruu) [inline .travis.yml]: Can you make travis config in a similar way as skt has? We are doing the installation of skt there directly, without mentioning the requirements, to ensure it works, and also testing uninstalling.

**Comment by danghai** [comment]
@veruu Thanks feedback, I fix the issue, and I also add pip install, and run tests instruction in `README`. You can review it.

File: README.md [hunk 4]
```diff
@@ -69,6 +69,22 @@ Install Kerberos client to enable `skt` authentication to Beaker:
 Install `skt` dependencies following its
 [README.md](https://github.com/RH-FMK/skt).
 
+Run tests
+---------
+
+For running all tests write down:
+
+    python -m unittest discover tests
```
Comment (veruu) [inline README.md]: We are using `pytest` for sktm, so both this and the line below should be modified to reflect on that.

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+[testenv]
+passenv = TRAVIS TRAVIS_*
+deps =
+    coverage
+    mock
+    coveralls
+    pytest
+commands =
+    coverage run --branch --source=sktm -m unittest discover tests
```
Comment (veruu) [inline tox.ini]: This should be using `pytest` as well.

File: README.md [hunk 6]
```diff
@@ -69,6 +69,22 @@ Install Kerberos client to enable `skt` authentication to Beaker:
...
+
+For running some specific tests you can do this as following:
+
+    python -m unittest tests.test_patchwork
+
+Installation
+------------
+
+    pip install git+https://github.com/rh-fmk/sktm
```
Comment (veruu) [inline README.md]: Can we use capital letters for `RH-FMK` here please?

File: README.md [hunk 7]
```diff
@@ -69,6 +69,22 @@ Install Kerberos client to enable `skt` authentication to Beaker:
...
+
+For running all tests write down:
+
+    python -m unittest discover tests
+
+For running some specific tests you can do this as following:
+
+    python -m unittest tests.test_patchwork
+
+Installation
```
Comment (veruu) [inline README.md]: Can you also mention the `-e` option for developer modes, as skt has?

**Comment by danghai** [comment]
@veruu you mean you want to use `sktm` instead of `sktm.py`?

**Comment by veruu** [comment]
@danghai not sure what you mean by it. One of the reasons to have a project installable is to be able to use the executable instead of calling the script manually.

**Comment by danghai** [comment]
@veruu I mean it relates to this issue #10  ?

**Comment by veruu** [comment]
I don't think so. That issue was created before we even thought about making the project installable, and it's mutually exclusive with it. Unless @spbnick who created it thinks otherwise, I'd close that issue if we are going with the installable.

**Comment by danghai** [comment]
Okay, anyway, I know what you mean. I will fix it now.

**Comment by danghai** [comment]
@veruu I fixed it. I hope it is better. It seems to me the `pip install/uninstall` does not work. Could you review?

File: README.md [hunk 8]
```diff
@@ -69,6 +69,22 @@ Install Kerberos client to enable `skt` authentication to Beaker:
...
+Run tests
+---------
+
+For running all tests write down:
+
+    python -m pytest -v -s tests
+
+For running some specific tests you can do this as following:
+
+    python -m pytest -v -s tests/test_patchwork
```
Comment (veruu) [inline README.md]: This line complains about nonexistent file. Can you fix it so we have examples that work?

File: README.md [hunk 9]
```diff
@@ -252,10 +268,10 @@ Create the `~/.sktmrc` file telling sktm how to access Jenkins:
     EOF
 
 Clone sktm with Git, or simply download and unpack the latest "master" branch
-archive. After that, you will be able to run the `sktm.py` executable
-directly. The examples below assume you run it while being in the source root
-directory, but any directory will do as long as the path to `sktm.py` is
-correct.
+archive. installs `sktm` by `pip install .`. After that, you will be able to
+run the `sktm` executable directly. The examples below assume you run it while
+being in the source root directory, but any directory will do as long as the
```
Comment (veruu) [inline README.md]: This doesn't matter as long as the installation completed successfully, so mentioning it is confusing. Can you remove this part? Actually, the whole paragraph can be removed since we have the installation section.

File: setup.cfg [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+                   enum34
+                   jenkinsapi
+                   junit_xml
+                   requests
+                   pylint
+                   mock
+
+[options.entry_points]
+console_scripts =
+    sktm = sktm.executable:main
```
Comment (veruu) [inline setup.cfg]: This change is done in two steps: in one commit, the `sktm:main` part is added, and then it's changed to include the `executable`. Can you do both in a single step?

**Comment by veruu** [comment]
Also, can you provide a `MANIFEST.in` with a list of files as skt has?

File: README.md (hunk 8 above)
Comment (danghai) [reply README.md]: I see it works in my environment. Are you sure you have `pytest` dependency?

Comment (veruu) [reply README.md]: Yes I am. There is no `test_patchwork` file in the repository.

Comment (danghai) [reply README.md]: `python -m pytest -v -s tests/test_patchwork.py`. Does it work with this command? I am not sure why it works with `test_patchwork` without `.py` in my environment

**Comment by danghai** [comment]
@veruu Thanks review. I fix these thing that you recommend.

**Comment by veruu** [comment]
@spbnick thanks for the comments! You raise a good point about the separation of the commits. However, I'd argue that the `.pytest_cache` one should be separate since it's unrelated to the rest (we are already using pytest in sktm and the directory is being created on test runs).

@danghai , can you merge the installation-related commits into one, as Nick suggests?

**Comment by spbnick** [comment]
Ah, yes, then the `.pytest_cache` one should be separate indeed.

**Comment by danghai** [comment]
@spbnick Thanks Nick. 
@veruu It means I need to merge the commit:
1. `Rename sktm.py to executabe.py, and add main() function`
2. `Make sktm install via pip`
3. `README.md: Add pip install, and run tests instructions`
4. `Add MANIFEST.in`
merge above these commits into a commit with name: `Make sktm install via pip`? I am little bit confused about `documentation`. It means the change in `README` or explanation in commit message? Thanks

**Comment by veruu** [comment]
@danghai yes, documentation are the README changes.

**Comment by danghai** [comment]
@veruu . I merge them into a commit. Could you re-check, and let me know if I need to change anything?

File: MANIFEST.in [hunk 11]
```diff
@@ -0,0 +1,4 @@
+include README.md
+include LICENSE
+include sktm/*
+include tests/*
```
Comment (veruu) [inline MANIFEST.in]: Can you also add the `db_migrations` directory here?

File: README.md [hunk 12]
```diff
@@ -69,6 +69,22 @@ Install Kerberos client to enable `skt` authentication to Beaker:
 Install `skt` dependencies following its
 [README.md](https://github.com/RH-FMK/skt).
 
+Run tests
```
Comment (veruu) [inline README.md]: Going with @spbnick 's comment about standalone commits, the section about test runs should be added in the commit that deals with tests.

File: README.md (hunk 7 above)
Comment (danghai) [reply README.md]: Do you mean below `Developer Guide`?

Comment (veruu) [reply README.md]: This is an old comment, you already resolved it with the developer guide

Comment (veruu) [reply README.md]: If the line of the code that was commented on doesn't change, the comment stays on github, so you have to check the date and time when figuring out what comments are new

Comment (danghai) [reply README.md]: Oh I see.

File: README.md (hunk 12 above)
Comment (danghai) [reply README.md]: Do I need to squash the commit `add .pytest_cache to gitignore` and `Add run tests section` into one commit?

Comment (veruu) [reply README.md]: No, those two changes are unrelated to each other.

Comment (danghai) [reply README.md]: Okay. could you review now?

File: .travis.yml [hunk 13]
```diff
@@ -4,11 +4,9 @@ python:
...
-    - pip install -U pip
-    - pip install -r requirements.txt
-    - pip install -r test-requirements.txt
-    - pip install tox
+    # Install sktm using pip to ensure dependencies are downloaded correctly.
+    - pip install .
+    - pip install coveralls tox
 script:
-    # Make sure all Python files are loadable
-    - find -name '*.py' | xargs -l1 -IFILE python -c 'import imp; imp.load_source("", "FILE")'
```
Comment (spbnick) [inline .travis.yml]: This change is unrelated to this commit and needs to be separate.

File: .travis.yml [hunk 14]
```diff
@@ -4,11 +4,9 @@ python:
...
     - tox -e $TOX_ENV
+    - pip uninstall -y sktm
```
Comment (spbnick) [inline .travis.yml]: Strictly speaking, this is also unrelated.

File: README.md [hunk 15]
```diff
@@ -69,6 +69,17 @@ Install Kerberos client to enable `skt` authentication to Beaker:
...
+---------
+
+For running all tests write down:
```
Comment (spbnick) [inline README.md]: "write down" would mean e.g. "write something down on a piece of paper". I.e. the reader is requested to make a note of the following command somewhere else. That will certainly not run the tests :) Personally, I would write something like "To run all tests execute:". This can be fixed in a separate PR.

File: README.md [hunk 16]
```diff
@@ -69,6 +69,17 @@ Install Kerberos client to enable `skt` authentication to Beaker:
...
+
+    python -m pytest -v -s tests
+
+For running some specific tests you can do this as following:
```
Comment (spbnick) [inline README.md]: This needs to say either "you can do the following" or "you can do as follows". I would write this as "You can execute a specific test like this:". This can also be fixed in a separate PR.

File: README.md (hunk 15 above)
Comment (veruu) [reply README.md]: I think the wording here and below was copied over from skt, so it will need to be changed in both :) That said, I agree with the comments.

**Comment by danghai** [comment]
@spbnick @veruu Thanks review. I fix it. I am not sure why the my order commit in local machine is different the order commit on PR. It seems the order commit is organized by time. In particular, the commit `Remove loadable command in Travis` is on the top of order, and the commit `Add check uninstall sktm command in Travis` is between 2nd and 3rd commit.

**Comment by veruu** [comment]
Merging since Nick approved this :) Thanks Hai!
